// stencil-apollo/dist/collection/components/apollo-subscription/apollo-subscription.js
import { ApolloProviderConsumer } from "../../utils/apollo-client-state";
export class ApolloSubscriptionComponent {
    componentWillLoad() {
        this.result = {
            data: undefined,
            errors: [],
            loading: true,
            networkStatus: undefined,
            stale: undefined
        };
        this.startSubscription();
    }
    onPropsChange() {
        this.stopSubscription();
        this.startSubscription();
    }
    componentDidUnload() {
        this.stopSubscription();
    }
    startSubscription() {
        if (this.client) {
            this._subscription = this.client.subscribe(Object.assign({ query: this.subscription, variables: this.variables }, this.options)).subscribe(result => {
                this.result = result;
                this.resultEventEmitter.emit(this.result);
            });
            this.readyEventEmitter.emit(this.result);
        }
        else {
            throw new Error('You should wrap your parent component with apollo-provider custom element or ApolloProvider functional component');
        }
    }
    stopSubscription() {
        if (this._subscription) {
            this._subscription.unsubscribe();
        }
    }
    render() {
        return this.renderer && this.renderer(this.result);
    }
    static get is() { return "apollo-subscription"; }
    static get properties() { return {
        "subscription": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "DocumentNode",
                "resolved": "any",
                "references": {
                    "DocumentNode": {
                        "location": "import",
                        "path": "graphql"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "subscription",
            "reflect": false
        },
        "renderer": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "SubscriptionRenderer<any>",
                "resolved": "(result: any) => Element | Element[]",
                "references": {
                    "SubscriptionRenderer": {
                        "location": "import",
                        "path": "../../utils/types"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "variables": {
            "type": "any",
            "mutable": false,
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            },
            "attribute": "variables",
            "reflect": false
        },
        "options": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "SubscriptionOptions",
                "resolved": "SubscriptionOptions<OperationVariables>",
                "references": {
                    "SubscriptionOptions": {
                        "location": "import",
                        "path": "apollo-client"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        },
        "client": {
            "type": "unknown",
            "mutable": false,
            "complexType": {
                "original": "ApolloClient<any>",
                "resolved": "ApolloClient<any>",
                "references": {
                    "ApolloClient": {
                        "location": "import",
                        "path": "apollo-client"
                    }
                }
            },
            "required": false,
            "optional": false,
            "docs": {
                "tags": [],
                "text": ""
            }
        }
    }; }
    static get states() { return {
        "result": {}
    }; }
    static get events() { return [{
            "method": "readyEventEmitter",
            "name": "ready",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }, {
            "method": "resultEventEmitter",
            "name": "result",
            "bubbles": true,
            "cancelable": true,
            "composed": true,
            "docs": {
                "tags": [],
                "text": ""
            },
            "complexType": {
                "original": "any",
                "resolved": "any",
                "references": {}
            }
        }]; }
    static get elementRef() { return "el"; }
    static get watchers() { return [{
            "propName": "client",
            "methodName": "onPropsChange"
        }, {
            "propName": "subscription",
            "methodName": "onPropsChange"
        }, {
            "propName": "variables",
            "methodName": "onPropsChange"
        }, {
            "propName": "renderer",
            "methodName": "onPropsChange"
        }, {
            "propName": "options",
            "methodName": "onPropsChange"
        }]; }
}
ApolloProviderConsumer.injectProps(ApolloSubscriptionComponent, ['client']);
