// stencil-apollo/dist/esm-es5/index.mjs
import { h } from './core-eaceb4a9.js';
import { A as ApolloProviderConsumer } from './apollo-client-state-7be4f0b5.js';
var Declarations = 'DECLARATIONS';
function ApolloProvider(_a, children) {
    var client = _a.client;
    return (h("apollo-provider", { client: client }, children));
}
function ApolloConsumer(_props, children) {
    return (h(ApolloProviderConsumer.Consumer, null, function (_a) {
        var client = _a.client;
        return children[0](client);
    }));
}
var Query = function (_a, children) {
    var query = _a.query, variables = _a.variables, options = _a.options;
    return h("apollo-query", { query: query, variables: variables, options: options, renderer: children[0] });
};
var Mutation = function (_a, children) {
    var mutation = _a.mutation, variables = _a.variables, options = _a.options;
    return h("apollo-mutation", { mutation: mutation, variables: variables, options: options, renderer: children[0] });
};
var Subscription = function (_a, children) {
    var subscription = _a.subscription, variables = _a.variables, options = _a.options;
    return h("apollo-subscription", { subscription: subscription, variables: variables, options: options, renderer: children[0] });
};
export { ApolloConsumer, ApolloProvider, Declarations, Mutation, Query, Subscription };
