// stencil-apollo/dist/esm/index.mjs
import { h } from './core-eaceb4a9.js';
import { A as ApolloProviderConsumer } from './apollo-client-state-7be4f0b5.js';

const Declarations = 'DECLARATIONS';

function ApolloProvider({ client }, children) {
    return (h("apollo-provider", { client: client }, children));
}
function ApolloConsumer(_props, children) {
    return (h(ApolloProviderConsumer.Consumer, null, ({ client }) => children[0](client)));
}

const Query = ({ query, variables, options, }, children) => h("apollo-query", { query: query, variables: variables, options: options, renderer: children[0] });

const Mutation = ({ mutation, variables, options }, children) => h("apollo-mutation", { mutation: mutation, variables: variables, options: options, renderer: children[0] });

const Subscription = ({ subscription, variables, options, }, children) => h("apollo-subscription", { subscription: subscription, variables: variables, options: options, renderer: children[0] });

export { ApolloConsumer, ApolloProvider, Declarations, Mutation, Query, Subscription };
