// stencil-apollo/dist/cjs/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const core = require('./core-4f654165.js');
const apolloClientState = require('./apollo-client-state-fb71bbf9.js');

const Declarations = 'DECLARATIONS';

function ApolloProvider({ client }, children) {
    return (core.h("apollo-provider", { client: client }, children));
}
function ApolloConsumer(_props, children) {
    return (core.h(apolloClientState.ApolloProviderConsumer.Consumer, null, ({ client }) => children[0](client)));
}

const Query = ({ query, variables, options, }, children) => core.h("apollo-query", { query: query, variables: variables, options: options, renderer: children[0] });

const Mutation = ({ mutation, variables, options }, children) => core.h("apollo-mutation", { mutation: mutation, variables: variables, options: options, renderer: children[0] });

const Subscription = ({ subscription, variables, options, }, children) => core.h("apollo-subscription", { subscription: subscription, variables: variables, options: options, renderer: children[0] });

exports.ApolloConsumer = ApolloConsumer;
exports.ApolloProvider = ApolloProvider;
exports.Declarations = Declarations;
exports.Mutation = Mutation;
exports.Query = Query;
exports.Subscription = Subscription;
